nvflare/private/fed/app/simulator/simulator_runner.py

```diff
@@ -569,9 +576,26 @@ def run_client_thread(self, num_of_threads, gpu, lock, rank, timeout=60):
                         self.end_run_clients.append(end_run_client)
 
                 client.simulate_running = False
+
+            if end_run_for_all:
+                self._end_run_clients(client, gpu, lock, num_of_threads, timeout)
         except Exception as e:
             self.logger.error(f"run_client_thread error: {secure_format_exception(e)}")
 
+    def _end_run_clients(self, client, gpu, lock, num_of_threads, timeout):
```

Comment:
What does this method do? I read this code many times and just couldn't understand it.
Can you rewrite it and add comments to make it easier to understand?

Reply:
Added the explanation of the execution logic. It's multiple threads pick up one of the "free resource" for execution.

Reply:
rewrote the _end_run_clients() for easier understanding.